presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Merge together the functions in WindowNodes that have identical WindowNode.Specifications.
+ * For example:
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--WindowNode(Specification: A, Functions: [avg(something)])
+ *             `--...
+ *
+ * Will be transformed into
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something), avg(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--...
+ *
+ * This will NOT merge the functions in WindowNodes that have identical WindowNode.Specifications,
+ * but have a node between them that is not a WindowNode.
+ * In the following example, the functions in the WindowNodes with specification `A' will not be
+ * merged into a single WindowNode.
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
+ *                `--...
+ */
+public class MergeIdenticalWindows
+        implements PlanOptimizer
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<MergeIdenticalWindows.Context>
+    {
+        @Override
+        public PlanNode visitWindow(
+                WindowNode node,
+                RewriteContext<MergeIdenticalWindows.Context> context)
+        {
+            Context mergeContext = context.get();
+            boolean retainNode = mergeContext.collectFunctions(node);
+
+            // Don't merge window functions across nodes that aren't WindowNodes
+            if (node.getClass().isInstance(node.getSource())) {
```

Comment:
Also I think this whole chunk might be a little simpler overall if we just:
1) collected all the window nodes as we moved down the plan (indexed by specification)
2) after hitting the first non-window node, just reconstructing all the necessary windows at the one place. we can use a linkedhashmap if order matters.
What do you think?